src/btree/row_modify.c

```diff
@@ -148,24 +148,30 @@ __wt_row_modify(WT_CURSOR_BTREE *cbt, const WT_ITEM *key, const WT_ITEM *value,
             __wt_upd_value_assign(cbt->modify_update, upd);
         } else {
             /*
-             * We only update history store records in three cases:
+             * We only update history store records in two cases:
              *  1) Delete the record with a tombstone with WT_TS_NONE.
-             *  2) Update the record's stop time point if the prepared update written to the data
-             * store is committed.
-             *  3) Reinsert an update that has been deleted by a prepared rollback.
+             *  2) Reinsert an update that has been deleted by a prepared commit or rollback.
              */
+            WT_ASSERT(session,
+              !WT_IS_HS(S2BT(session)->dhandle) || *upd_entry == NULL ||
+                (*upd_entry)->next == NULL ||
+                ((*upd_entry)->type == WT_UPDATE_STANDARD &&
+                  (*upd_entry)->txnid != WT_TXN_ABORTED) ||
+                ((*upd_entry)->type == WT_UPDATE_TOMBSTONE && (*upd_entry)->txnid == WT_TXN_NONE &&
+                  (*upd_entry)->upd_start_ts == WT_TS_NONE) ||
+                ((*upd_entry)->txnid == WT_TXN_ABORTED &&
+                  (*upd_entry)->next->txnid == WT_TXN_ABORTED) ||
+                ((*upd_entry)->type == WT_UPDATE_TOMBSTONE &&
+                  (*upd_entry)->txnid != WT_TXN_ABORTED &&
+                  (*upd_entry)->next->type == WT_UPDATE_STANDARD &&
+                  (*upd_entry)->next->txnid != WT_TXN_ABORTED));
             WT_ASSERT(session,
               !WT_IS_HS(S2BT(session)->dhandle) ||
-                (*upd_entry == NULL ||
-                  ((*upd_entry)->type == WT_UPDATE_TOMBSTONE &&
-                    (((*upd_entry)->txnid == WT_TXN_NONE &&
-                       (*upd_entry)->upd_start_ts == WT_TS_NONE) ||
-                      ((*upd_entry)->txnid == WT_TXN_ABORTED &&
-                        (*upd_entry)->next->txnid == WT_TXN_ABORTED)))) ||
                 (upd_arg->type == WT_UPDATE_TOMBSTONE && upd_arg->upd_start_ts == WT_TS_NONE &&
                   upd_arg->next == NULL) ||
                 (upd_arg->type == WT_UPDATE_TOMBSTONE && upd_arg->next != NULL &&
-                  upd_arg->next->type == WT_UPDATE_STANDARD && upd_arg->next->next == NULL));
+                  upd_arg->next->type == WT_UPDATE_STANDARD && upd_arg->next->next == NULL) ||
+                (upd_arg->type == WT_UPDATE_STANDARD && upd_arg->next == NULL));
```

Comment:
added this case to assertion, since we can also insert a standard update to the history store